### ISSUE TITLE
Platform Services Initialization failed

### ISSUE THREAD

**Issue by yarnauy** [apache/teaclave-sgx-sdk#3]
I started the docker with "--device /dev/isgx" , but when I ran  "/opt/intel/sgxpsw/aesm/aesm_service" , it failed:
```shell
root@ae6fcae98e03:/# /opt/intel/sgxpsw/aesm/aesm_service 
root@ae6fcae98e03:/# aesm_service[17]: [ADMIN]Platform Services initializing
aesm_service[17]: [ADMIN]Platform Services initialization failed due to DAL error
aesm_service[17]: The server sock is 0xe1d5b0
```

My host:
Ubuntu 16.04
Intel SGX SDK **1.9** for Linux
rustc 1.19.0-nightly (386b0b9d3 2017-05-14)

Actually, I have installed sgx-sdk and ran the samplecode in sgx-sdk successfully on my host.

So I tried to set up the environment directly on my host (**refer to your dockerfile** ) but failed again. I can compile your samplecode in the docker but failed on my host as below.
```shell
yr@yr-desktop:~/Workspace/rust-sgx-sdk/samplecode/sealeddata$ make
/opt/intel/sgxsdk/bin/x64/sgx_edger8r --trusted enclave/Enclave.edl --search-path /opt/intel/sgxsdk/include --trusted-dir enclave
/opt/intel/sgxsdk/bin/x64/sgx_edger8r --untrusted enclave/Enclave.edl --search-path /opt/intel/sgxsdk/include --untrusted-dir app
GEN  =>  enclave/Enclave_t.c enclave/Enclave_t.h app/Enclave_u.c app/Enclave_u.h
CC   <=  app/Enclave_u.c
CXX  <=  app/app.c
LINK =>  bin/app
make -C ./enclave/
make[1]: Entering directory '/home/yr/Workspace/rust-sgx-sdk/samplecode/sealeddata/enclave'
cargo build --release
   Compiling sgx_types v0.1.0 (file:///home/yr/Workspace/rust-sgx-sdk/sgx_types)
   Compiling sgx_trts v0.1.0 (file:///home/yr/Workspace/rust-sgx-sdk/sgx_trts)
   Compiling sgx_tcrypto v0.1.0 (file:///home/yr/Workspace/rust-sgx-sdk/sgx_tcrypto)
   Compiling sgx_tse v0.1.0 (file:///home/yr/Workspace/rust-sgx-sdk/sgx_tse)
   Compiling sgx_tseal v0.1.0 (file:///home/yr/Workspace/rust-sgx-sdk/sgx_tseal)
   Compiling Seal data sample enclave v0.1.0 (file:///home/yr/Workspace/rust-sgx-sdk/samplecode/sealeddata/enclave)
note: link against the following native artifacts when linking against this static library

note: the order and any duplication can be significant on some platforms, and so may need to be preserved

note: library: sgx_tstdc

note: library: sgx_tservice

note: library: sgx_tcrypto

note: library: sgx_tkey_exchange

note: library: sgx_trts

note: library: sgx_uae_service

note: library: sgx_ukey_exchange

note: library: sgx_urts

note: library: c

note: library: m

note: library: rt

note: library: pthread

note: library: util

    Finished release [optimized] target(s) in 1.13 secs
make[1]: Leaving directory '/home/yr/Workspace/rust-sgx-sdk/samplecode/sealeddata/enclave'
make -C ../../compiler-rt/ 2> /dev/null
make[1]: Entering directory '/home/yr/Workspace/rust-sgx-sdk/compiler-rt'
make[1]: Nothing to be done for 'all'.
make[1]: Leaving directory '/home/yr/Workspace/rust-sgx-sdk/compiler-rt'
CC   <=  enclave/Enclave_t.c
cp ./enclave/target/release/libenclave.a ./lib
cp ../../compiler-rt/libcompiler-rt-patch.a ./lib
./lib/libenclave.a(alloc_system-06d737faf6076d7c.0.o): In function `alloc_system::imp::aligned_malloc':
/checkout/src/liballoc_system/lib.rs:121: undefined reference to `posix_memalign'
/checkout/src/liballoc_system/lib.rs:121: undefined reference to `posix_memalign'
/checkout/src/liballoc_system/lib.rs:121: undefined reference to `posix_memalign'
collect2: error: ld returned 1 exit status
Makefile:159: recipe for target 'enclave/enclave.so' failed
make: *** [enclave/enclave.so] Error 1

```

How can I fix them...

**Comment by dingelish** [comment]
For the first problem, the reason is that the driver in SGX v1.9 is different from v1.8. You are using the v1.8 docker which contains v1.8 platform service along with v1.8 Intel SDK and they cannot work well with v1.9 driver.

For the second problem, pay attention to the [patch](https://github.com/baidu/rust-sgx-sdk/blob/master/dockerfile/patch) file. If you are using v1.9 toolchain, please install v1.9 driver first, and then the platform service as well as the SDK from 01.org. Then checkout the github repo from 01.org and patch sdk/tlibc/Makefile with the same patch. Then `make` and copy libsgx_tstdc.a to /opt/sgxsdk/lib64/libsgx_tstdc.a and you'll be fine.

Great news is that we are releasing v0.2.0 this weekend. You could see that the thread sample case may not work in SGX v1.9 and v0.2.0 could fix this. What's more, v0.2.0 supports a bunch of threading features and more than 4GB enclave memory. I'm finalizing some docs and will release it very soon.

**Comment by njuyuanrui** [comment]
thank you

**Comment by yarnauy** [comment]
thank you @dingelish

**Comment by Pratyush** [comment]
I get the PSI error with the new SDK too:

```
 /opt/intel/sgxpsw/aesm/aesm_service &
[1] 17
root@79d2b9eb1423:~# aesm_service[18]: [ADMIN]White List update requested
aesm_service[18]: The server sock is 0x2438530
aesm_service[18]: [ADMIN]Platform Services initializing
aesm_service[18]: [ADMIN]Platform Services initialization failed due to DAL error
aesm_service[18]: [ADMIN]White list update request successful for Version: 21
```
Subsequently, when I try to compile and run the enclave, I get the following error:

```
root@79d2b9eb1423:~/sgx/samplecode/hello-rust# ./bin/app
[+] Home dir is /root
[-] Open token file /root/enclave.token error! Will create one.
[-] Init Enclave Failed SGX_ERROR_ENCLAVE_FILE_ACCESS!
```

**Comment by dingelish** [comment]
Hi Pratyush,

Did you use docker argument `--device /dev/isgx` to get `/dev/isgx` works inside the docker container? Thanks!

Best,
Yu

**Comment by Pratyush** [comment]
Yes, I used the command in the README. What's weird is that the sample code was working initially. Then I tried to write an enclave of my own, and since then it's not been working. On the other hand, if I use Jethro's SDK (https://github.com/jethrogb/sgx-utils), my enclave runs fine.

I've tried removing the docker image, redownloading the rust-sdk, and rerunning the examples, but to no avail. I've also reinstalled the latest SGX driver multiple times.

**Comment by dingelish** [comment]
This is really weird. I've been working in docker container on this SDK for thousands of hours without such error. 

Could you please double check the driver is v1.9?

btw, do you mind providing a test account on your machine and I can debug it for you remotely? Thanks!

You can write email to me in private at dingelish gmail

**Comment by fredfortier** [comment]
I having the same error with `baiduxlab/sgx-rust:1804` and v1.0.4. Here is my host kernel: `4.19.15-300.fc29.x86_64+debug`. I'm currently compiling `https://github.com/intel/linux-sgx-driver` from this commit: `b73806087e14bf16eceb6f96f2689a459aa56144` (because it worked for me before). The driver seems to work fine on my host. I can even run build and run all of the `samplecode` packages that I tried on the Docker image.

This used to work perfectly for me. This issue started when I upgraded my host kernel to 4.20, realized that it wasn't compatible with the SGX linux driver (they fixed that since) and downgraded back 4.19. I can't easily upgrade to v1.0.5. Any troubleshooting pointers would be appreciated.
